sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowJson.java

```diff
@@ -0,0 +1,365 @@
+      super(Row.class);
+      this.schema = schema;
+    }
+
+    @Override
+    public Row deserialize(JsonParser jsonParser, DeserializationContext deserializationContext)
+        throws IOException {
+
+      // Parse and convert the root object to Row as if it's a nested field with name 'root'
+      return (Row)
+          extractJsonNodeValue(
+              FieldValue.of("root", FieldType.row(schema), jsonParser.readValueAsTree()));
+    }
+
+    private static Object extractJsonNodeValue(FieldValue fieldValue) {
+      if (!fieldValue.isJsonValuePresent()) {
+        throw new UnsupportedRowJsonException(
+            "Field '" + fieldValue.name() + "' is not present in the JSON object");
+      }
+
+      if (fieldValue.isJsonNull()) {
+        return null;
+      }
+
+      if (fieldValue.isRowType()) {
+        return jsonObjectToRow(fieldValue);
+      }
+
+      if (fieldValue.isArrayType()) {
+        return jsonArrayToList(fieldValue);
+      }
+
+      return extractJsonPrimitiveValue(fieldValue);
+    }
+
+    private static Row jsonObjectToRow(FieldValue rowFieldValue) {
+      if (!rowFieldValue.isJsonObject()) {
+        throw new UnsupportedRowJsonException(
+            "Expected JSON object for field '"
+                + rowFieldValue.name()
+                + "'. "
+                + "Unable to convert '"
+                + rowFieldValue.jsonValue().asText()
+                + "'"
+                + " to Beam Row, it is not a JSON object. Currently only JSON objects "
+                + "can be parsed to Beam Rows");
+      }
+
+      return rowFieldValue.rowSchema().getFields().stream()
+          .map(
+              schemaField ->
+                  extractJsonNodeValue(
+                      FieldValue.of(
+                          schemaField.getName(),
+                          schemaField.getType(),
+                          rowFieldValue.jsonFieldValue(schemaField.getName()))))
+          .collect(toRow(rowFieldValue.rowSchema()));
+    }
+
+    private static Object jsonArrayToList(FieldValue arrayFieldValue) {
+      if (!arrayFieldValue.isJsonArray()) {
+        throw new UnsupportedRowJsonException(
+            "Expected JSON array for field '"
+                + arrayFieldValue.name()
+                + "'. "
+                + "Instead got "
+                + arrayFieldValue.jsonNodeType().name());
+      }
+
+      return arrayFieldValue
+          .jsonArrayElements()
+          .map(
+              jsonArrayElement ->
+                  extractJsonNodeValue(
+                      FieldValue.of(
+                          arrayFieldValue.name() + "[]",
+                          arrayFieldValue.arrayElementType(),
+                          jsonArrayElement)))
+          .collect(toList());
+    }
+
+    private static Object extractJsonPrimitiveValue(FieldValue fieldValue) {
+      try {
+        return JSON_VALUE_GETTERS.get(fieldValue.typeName()).extractValue(fieldValue.jsonValue());
```

Comment:
nit: this might be more readable as a switch statement, but it looks like you are just moving it.

Reply:
Yeah agreed a switch would be better. Do you want me to change that for this PR?

Reply:
It is fine as is.